04-cs-fundamentals/classroom/recursion.md

```diff
@@ -0,0 +1,343 @@
+
+<details>
+  <summary>What do you think the space complexity of factorial is?</summary>
+  O(n)
+</details>
+
+### Tracing Through Factorial
+
+If you have `factorial(5)` first this method call gets put on the call stack.
+
+![Factorial(5)](images/factorial-5.png)
+
+Then `factorial(5)` calls `factorial(4)`, which calls `factorial(3)`, and that calls `factorial(2)` and this calls `factorial(1)`.
+
+![Factorial 5 -> 1 ](images/factorial5-1.png)
+
+`factorial(1)` returns 1, which is plugged into `factorial(2)` which multiplies 2 * that returned result and carries on down in this fashion.
+
+![factorial(5) completely executed](images/factorial-5-complete.png)
+
+So you can trace a recursive function by walking through it's function calls, placing them on a simulated stack and then unwinding the stack as methods complete returning their result to the calling method.  
+
+
+### Understanding space and time complexities for the example
+
+To compute *factorial(n)*, n operations will be needed. Therefore, the time complexity will be *O(n)*.</br>
+Each recursive call will end up with a stack frame on the call stack. There will be *n* such stack frames by the time the base case is reached and the stack starts unwinding. Stack frames take up space in memory. So, the space complexity will be *O(n)*.
+
+### Writing A Recursive Method
+
+When writing a recursive method, try the following:
+
+- Think about the recursive case: How are we going to break the problem into a smaller problem (by calling out method again) and get closer to the base case?
+- Think about how you can reduce the problem to one or more smaller sub-problems of the same form.
+- Think about what information you need to give to the sub-problems (the parameters).
+- Think about what information you want back from the sub-problems (the return type).
+- Write the method header.
+- Think about the base case: When is the answer so simple that we know the answer without recursing?
+- Write a method specification (like the static view of the problem) that explains exactly what it will do in terms of the parameters. Include any preconditions.
+- Write the code.
+- Test out your code with several different cases. Ensure all of them terminate with a base case and yield the right results.
+
+## Example - Solving 
+
+Consider this programming problem.
+
+```
+Devise an algorithm for a function that takes a natural number as input parameter and computes the sum of all natural number up to and including the input parameter.
+
+E.g. 
+natural_numbers_sum(3) should return 6 = 3 + 2 + 1
+natural_numbers_sum(4) should return 10 = 4 + 3 + 2 + 1
+```
+
+Note:
+
+In mathematics, the natural numbers are those used for counting and ordering. Natural numbers start with 0 or 1. (aka positive integers).
+
+You could write an iterative solution like this:
+
+```ruby
```

Comment:
It might be worth hiding the code inside a `<describe>` .

Reply:
The problem is I can't put a code block inside a `describe`  tag.

Reply:
You can but Github is _very_ particular about how you format it.